tululbot/utils.py

```diff
@@ -0,0 +1,170 @@
+
+        def parse_first_paragraph(page):
+            return parse_content_text(page).find('p')
+
+        def has_result(page):
+            return ('Search results' not in page and
+                    parse_first_paragraph(page) is not None)
+
+        def has_disambiguations(paragraph):
+            return 'may refer to:' in paragraph
+
+        def parse_first_disambiguation_link(page):
+            def valid_link(tag):
+                return tag.name == 'a' and tag['href'].startswith('/wiki')
+
+            path = parse_content_text(page).find(valid_link)['href']
+            return 'https://en.wikipedia.org{}'.format(path)
+
+        query_string = urlencode(dict(search=term))
+        search_url = \
+            'https://en.wikipedia.org/w/index.php?{}'.format(query_string)
+
+        response = requests.get(search_url)
+        if response.status_code != 200:
+            return None
+
+        page = response.text
+        if not has_result(page):
+            return None
+
+        first_paragraph = parse_first_paragraph(page)
+        if not has_disambiguations(first_paragraph.get_text()):
+            return first_paragraph.get_text()
+
+        disambiguation_url = parse_first_disambiguation_link(page)
+
+        response = requests.get(disambiguation_url)
+        if response.status_code != 200:
+            return None
+
+        disambiguated_page = response.text
+        return parse_first_paragraph(disambiguated_page).get_text()
+
+    def search_on_google():
+        query_string = urlencode(dict(q=term))
+        search_url = 'https://google.com/search?{}'.format(query_string)
+        return 'Jangan ngeleli! Googling dong: {}'.format(search_url)
+
+    result = search_on_wikipedia()
+    return result if result is not None else search_on_google()
+
+
+class QuoteEngine:
+
+    def __init__(self):
+        self._quote_url = 'https://cdn.rawgit.com/tulul/tulul-quotes/master/quote.yaml'  # noqa
+        # Note: rawgit does not have 100% uptime, but at
+        # least they're not throttling us.
+
+        self._cache = []
+        self._cache_refreshed = False
+
+    def retrieve_random(self):
+        cache = self._cache
+        return self.format_quote(random.choice(cache))
+
+    def format_quote(self, q):
+        return '{q[quote]} - {q[author]}, {q[author_bio]}'.format(q=q)
+
+    def refresh_cache(self):
+        if not self._cache_refreshed:
+            body = requests.get(self._quote_url).text
+            # What if previosuly we have the cache, but this time
+            # when we try to get new cache, the network occurs error?
+            # We will think about "don't refresh if error" later.
+            self._cache = yaml.load(body)['quotes']
+            self._cache_refreshed = True
+
+
+class TululBot:
+    """Class representing a TululBot."""
+
+    def __init__(self):
+        self._command_list = []
+
+    def add_command(self, pattern, callback):
```

Comment:
If we need more params, just partially apply the function first. There are other ways too. Class is not necessary.